src/tools/rbd_mirror/ImageDeleter.cc

```diff
@@ -0,0 +1,367 @@
+                            retries);
+      return;
+    }
+
+    for (const auto& snap : snaps) {
+      dout(20) << "processing deletion of snapshot " << imgctx->name << "@"
+               << snap.name << dendl;
+
+      bool is_protected;
+      r = librbd::snap_is_protected(imgctx, snap.name.c_str(), &is_protected);
+      if (r < 0) {
+        derr << "error checking snapshot protection of snapshot "
+             << imgctx->name << "@" << snap.name << ": " << cpp_strerror(r)
+             << dendl;
+        imgctx->state->close();
+        enqueue_failed_delete(std::move(del_info), FAILED_ON_REMOVE_SNAPS, r,
+                              retries);
+        return;
+      }
+      if (is_protected) {
+        dout(20) << "snapshot " << imgctx->name << "@" << snap.name
+                 << " is protected, issuing unprotect command" << dendl;
+
+        r = imgctx->operations->snap_unprotect(snap.name.c_str());
+        if (r < 0) {
+          derr << "error unprotecting snapshot " << imgctx->name << "@"
+               << snap.name << ": " << cpp_strerror(r) << dendl;
+          imgctx->state->close();
+          enqueue_failed_delete(std::move(del_info), FAILED_ON_REMOVE_SNAPS, r,
+                                retries);
+          return;
+        }
+      }
+
+      r = imgctx->operations->snap_remove(snap.name.c_str());
+      if (r < 0) {
+        derr << "error removing snapshot " << imgctx->name << "@"
+             << snap.name << ": " << cpp_strerror(r) << dendl;
+        imgctx->state->close();
+        enqueue_failed_delete(std::move(del_info), FAILED_ON_REMOVE_SNAPS, r,
+                              retries);
+        return;
+      }
+
+      dout(20) << "snapshot " << imgctx->name << "@" << snap.name
+               << " was deleted" << dendl;
+
+    }
+
+    imgctx->state->close();
+  }
+
+failed_on_remove_image:
+  class SilentProgressContext : public ProgressContext {
+    int update_progress(uint64_t offset, uint64_t total) {
+      return 0;
+    }
+  } silent_progress;
+  r = librbd::remove(ioctx, img_name.c_str(), silent_progress, true);
+  if (r < 0) {
+    derr << "error removing image " << img_name << " from local pool: "
+         << cpp_strerror(r) << dendl;
+    enqueue_failed_delete(std::move(del_info), FAILED_ON_REMOVE_IMAGE, r,
+                          retries);
+    return;
+  }
+
+  dout(20) << "image " << img_name << " deleted successfully" << dendl;
+}
+
+int ImageDeleter::image_has_snapshots_and_children(IoCtx *ioctx,
+                                                   string& image_id,
+                                                   bool *has_snapshots,
+                                                   bool *has_children) {
+
+  *has_snapshots = false;
+  *has_children = false;
+
+  string header_oid = librbd::util::header_name(image_id);
+  ::SnapContext snapc;
+  int r = cls_client::get_snapcontext(ioctx, header_oid, &snapc);
```

Comment:
Never mind -- I see that you are doing that after invoking this helper function.